zephyr/include/sof/lib/regions_mm.h

```diff
@@ -10,31 +10,239 @@
+	       (uint8_t *)ptr < (uint8_t *)range_start + range_size;
+}
+
+/**
+ * @brief Checks if given memory range overlaps second range.
+ *
+ * @param checked_memory_ptr start of first checked memory range
+ * @param checked_memory_size size of first checked memory range.
+ * @param range_start start of memory range
+ * @param range_size size of range.
+ *
+ * @retval false if no overlap detected.
+ * @retval true if any overlap detected.
+ */
+static inline bool vmh_do_ranges_overlap(void *checked_memory_ptr, size_t checked_memory_size,
+					 void *range_start, size_t range_size)
+{
+	uint8_t *checked_memory_end = (uint8_t *)checked_memory_ptr + checked_memory_size;
+	uint8_t *range_end = (uint8_t *)range_start + range_size;
+
+	return ((uint8_t *)checked_memory_ptr >= (uint8_t *)range_start &&
+		(uint8_t *)checked_memory_ptr < range_end) ||
+		(checked_memory_end > (uint8_t *)range_start &&
+		checked_memory_end <= range_end);
+}
+
+/**
+ * @brief Checks if given memory range overlaps second range
+ * in full.
+ *
+ * @param checked_memory_ptr start of first checked memory range
+ * @param checked_memory_size size of first checked memory range.
+ * @param range_start start of range
+ * @param range_size size of memory range.
+ *
+ * @retval false if no overlap detected or checked memory is not
+ * fully encapsulated in range.
+ * @retval true if checked memory is fully in given range.
+ */
+static inline bool vmh_is_range_fully_in_range(void *checked_memory_ptr,
+					       size_t checked_memory_size, void *range_start,
+					       size_t range_size)
+{
+	uint8_t *checked_memory_end = (uint8_t *)checked_memory_ptr + checked_memory_size;
+	uint8_t *range_end = (uint8_t *)range_start + range_size;
+
+	return (uint8_t *)checked_memory_ptr >= (uint8_t *)range_start &&
+		(uint8_t *)checked_memory_ptr < range_end &&
+		checked_memory_end > (uint8_t *)range_start &&
+		checked_memory_end <= range_end;
+}
+
+/**
+ * @brief Saves all ptrs that are not already mapped from
+ * provided memory range.
+ *
+ * @param ptrs_to_map table of ptrs that should be mapped
+ * @param aligned_ptr start of range.
+ * @param aligned_alloc_size size of checked range
+ * @param max_pointers_to_map size of ptrs table
+ */
+static inline void vmh_get_ptrs_to_map(char **ptrs_to_map, void *aligned_ptr,
+				       size_t aligned_alloc_size, int max_pointers_to_map)
+{
+	struct list_item *mappings_iterator;
+	int i;
+
+	for (i = 0; i < max_pointers_to_map; i++) {
+		bool need_mapping = true;
+		int offset = i * CONFIG_MM_DRV_PAGE_SIZE;
+
+		list_for_item(mappings_iterator, &vmh_active_physical_mappings_list) {
+			struct vmh_active_physical_mapping *mapping =
+				CONTAINER_OF(mappings_iterator,
+					struct vmh_active_physical_mapping, node);
+
+			need_mapping = vmh_is_ptr_in_memory_range(mapping->mapping_ptr,
+				(void *)((char *)aligned_ptr + offset),
+			    CONFIG_MM_DRV_PAGE_SIZE);
+			if (need_mapping)
+				continue;
```

Comment:
ehm, these two lines are a NOP here. Did you mean to do `if (!need_mapping) break;`?